lib/cartopy/crs.py

```diff
@@ -10,23 +10,619 @@
+
+    def __init__(self, proj4_params, globe=None):
+        """
+        Parameters
+        ----------
+        proj4_params: iterable of key-value pairs
+            The proj4 parameters required to define the
+            desired CRS.  The parameters should not describe
+            the desired elliptic model, instead create an
+            appropriate Globe instance. The ``proj4_params``
+            parameters will override any parameters that the
+            Globe defines.
+        globe: :class:`~cartopy.crs.Globe` instance, optional
+            If omitted, the default Globe instance will be created.
+            See :class:`~cartopy.crs.Globe` for details.
+
+        """
+        # for compatibility with pyproj.CRS and rasterio.crs.CRS
+        try:
+            proj4_params = proj4_params.to_wkt()
+        except AttributeError:
+            pass
+        # handle PROJ JSON
+        if (
+            isinstance(proj4_params, dict) and
+            "proj" not in proj4_params and
+            "init" not in proj4_params
+        ):
+            proj4_params = json.dumps(proj4_params)
+
+        if globe is not None and isinstance(proj4_params, str):
+            raise ValueError("Cannot have 'globe' with string params.")
+        if globe is None and not isinstance(proj4_params, str):
+            if self._handles_ellipses:
+                globe = Globe()
+            else:
+                globe = Globe(semimajor_axis=WGS84_SEMIMAJOR_AXIS,
+                              ellipse=None)
+        if globe is not None and not self._handles_ellipses:
+            a = globe.semimajor_axis or WGS84_SEMIMAJOR_AXIS
+            b = globe.semiminor_axis or a
+            if a != b or globe.ellipse is not None:
+                warnings.warn('The "{}" projection does not handle elliptical '
+                              'globes.'.format(self.__class__.__name__))
+        self.globe = globe
+        if isinstance(proj4_params, str):
+            self._proj4_params = {}
+            self.proj4_init = proj4_params
+        else:
+            self._proj4_params = self.globe.to_proj4_params()
+            self._proj4_params.update(proj4_params)
+
+            init_items = []
+            for k, v in self._proj4_params.items():
+                if v is not None:
+                    if isinstance(v, float):
+                        init_items.append('+{}={:.16}'.format(k, v))
+                    elif isinstance(v, np.float32):
+                        init_items.append('+{}={:.8}'.format(k, v))
+                    else:
+                        init_items.append('+{}={}'.format(k, v))
+                else:
+                    init_items.append('+{}'.format(k))
+            self.proj4_init = ' '.join(init_items) + ' +no_defs'
+        super().__init__(self.proj4_init)
+
+    def __eq__(self, other):
+        if isinstance(other, CRS):
+            # Fast path Cartopy's CRS
+            return self.proj4_init == other.proj4_init
```

Comment:
I think that you also need to check that this is not None.

Reply:
Actually, nevermind.

Reply:
Hm... I don't think it can be? Looking through the __init__ it looks like it has to be a string in both if branches.